rfcs/0049-flakes.md

```diff
@@ -0,0 +1,496 @@
+    fetched as tarballs rather than as Git repositories.
+
+  * `lastModifiedDate`: The commit time of the revision `rev`, in the
+    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
+    this is available for both Git and GitHub repositories, so it's
+    useful for generating (hopefully) monotonically increasing version
+    strings.
+
+  * `lastModified`: The commit time of the revision `rev` as an integer
+    denoting the number of seconds since 1970.
+
+  * `narHash`: The SHA-256 (in SRI format) of the NAR serialization of
+    the flake's source tree.
+
+  The value returned by the `outputs` function must be an attribute
+  set. The attributes can have arbitrary values; however, some tools
+  may require specific attributes to have a specific value (e.g. the
+  `nix` command may expect the value of `packages.x86_64-linux` to be
+  an attribute set of derivations built for the `x86_64-linux`
+  platform).
+
+## Flake inputs
+
+The attribute `inputs` specifies the dependencies of a flake. These
+specify the location of the dependency, or a symbolic flake identifier
+that is looked up in a registry or in a command-line flag. For
+example, the following specifies a dependency on the Nixpkgs and import-cargo
+repositories:
+
+    # A GitHub repository.
+    inputs.import-cargo = {
+      type = "github";
+      owner = "edolstra";
+      repo = "import-cargo";
+    };
+
+    # An indirection through the flake registry.
+    inputs.nixpkgs = {
+      type = "indirect";
+      id = "nixpkgs";
+    };
+
+Each input is fetched, evaluated and passed to the `outputs` function
+as a set of attributes with the same name as the corresponding
+input. The special input named `self` refers to the outputs and source
+tree of *this* flake. Thus, a typical `outputs` function looks like
+this:
+
+    outputs = { self, nixpkgs, import-cargo }: {
+      ... outputs ...
+    };
+
+It is also possible to omit inputs entirely and *only* list them as
+expected function arguments in `outputs`. Thus,
+
+    outputs = { self, nixpkgs }: ...;
+
+without an `inputs.nixpkgs` attribute will simply look up `nixpkgs` in
+the flake registry.
+
+Repositories that don't contain a `flake.nix` can also be used as
+inputs, by setting the input's `flake` attribute to `false`:
+
+    inputs.grcov = {
+      type = "github";
+      owner = "mozilla";
+      repo = "grcov";
+      flake = false;
+    };
+
+    outputs = { self, nixpkgs, grcov }: {
+      packages.x86_64-linux.grcov = stdenv.mkDerivation {
+        src = grcov;
+        ...
+      };
+    };
+
+The following input types are specified at present:
+
+* `git`: A Git repository or dirty local working tree.
+
+* `github`: A more efficient scheme to fetch repositories from GitHub
+  as tarballs. These have slightly different semantics from `git`
+  (in particular, the `revCount` attribute is not available).
+
+* `gitlab`: Like `github`, but for Git repositories hosted on GitLab.
```

Comment:
I don't think this should have been dropped in during the FCP, as it seems very underspecified. It doesn't clearly define the parameters that this input takes. GitLab can be self-hosted (so you can't assume the repo is at `gitlab.com`, is there a way to change that?), \*and\* it has a concept of (recursively nested) subgroups, which don't really play well with the `owner`/`repo` keys that `github` uses.